ide.lua

```diff
@@ -107,41 +121,42 @@ local function editor(aceEnabled) -- feel free to disable the shiny Ajax.org Clo
             elseif url:match(".json") then mode = mode .. 'json'
             elseif url:match(".js") then mode = mode .. 'javascript'
+            elseif url:match(".md") then mode = mode .. 'markdown'
             else mode = mode .. 'lua'
             end
             sen = sen .. "<style type='text/css'>#editor{width: 100%; height: 80%}</style><div id='editor'></div><script src='//rawgit.com/ajaxorg/ace-builds/master/src-min-noconflict/ace.js'></script>"
-                .. "<script>var e=ace.edit('editor');e.setTheme('ace/theme/monokai');e.getSession().setMode('"..mode.."');function getSource(){return e.getValue();};function setSource(s){e.setValue(s);}</script>"
+                 .. "<script>var e=ace.edit('editor');e.setTheme('ace/theme/monokai');e.getSession().setMode('"..mode.."');function getSource(){return e.getValue();};function setSource(s){e.setValue(s);}</script>"
         else
             sen = sen .. "<textarea name=t cols=79 rows=17></textarea></br>"
-                .. "<script>function getSource() {return document.getElementsByName('t')[0].value;};function setSource(s) {document.getElementsByName('t')[0].value = s;};</script>"
+                 .. "<script>function getSource() {return document.getElementsByName('t')[0].value;};function setSource(s) {document.getElementsByName('t')[0].value = s;};</script>"
         end
         sen = sen .. "<script>function tag(c){document.getElementsByTagName('w')[0].innerHTML=c};var x=new XMLHttpRequest();x.onreadystatechange=function(){if(x.readyState==4) setSource(x.responseText);};"
-	    .. "x.open('GET',location.pathname);x.send()</script><button onclick=\"tag('Saving, wait!');x.open('POST',location.pathname);x.onreadystatechange=function(){console.log(x.readyState);"
-	    .. "if(x.readyState==4) tag(x.responseText);};x.send(new Blob([getSource()],{type:'text/plain'}));\">Save</button> <a href='?run'>[Run File]</a> <a href=\"/\">[Main Page]</a> <w></w>"
+        .. "x.open('GET',location.pathname);x.send()</script><button onclick=\"tag('Saving, wait!');x.open('POST',location.pathname);x.onreadystatechange=function(){console.log(x.readyState);"
+        .. "if(x.readyState==4) tag(x.responseText);};x.send(new Blob([getSource()],{type:'text/plain'}));\">Save</button> <a href='?run'>[Run File]</a> <a href=\"/\">[Main Page]</a> <w></w>"
 
     elseif vars == "run" then
         sen = sen .. "Output of the run:<hr><pre>"
-
-        function s_output(str) sen = sen .. str end
-        node.output(s_output, 0) -- re-direct output to function s_output.
-
-        local st, result = pcall(dofile, url)
-
-        -- delay the output capture by 1000 milliseconds to give some time to the user routine in pcall()
-        tmr.alarm(0, 1000, tmr.ALARM_SINGLE, function() 
-            node.output(nil)
+        local function output_CB(opipe)   -- upval: sck
+           for line in opipe:reader() do
+             sen = sen .. line .. "\n"
```

Comment:
That's not really the way pipe is meant.
You should insted read the whole pipe at the end of the timer. Concating the string multiple times creates lots of short lived objects which have to be collected by GC. To avoid this the pipe module was introduced.

See https://github.com/nodemcu/nodemcu-firmware/blob/release/lua_modules/telnet/telnet.lua for an example